### PR TITLE
Setup and use DEAL_II_PETSC_WITH_MUMPS in PETSc solver.

### PR DESCRIPTION
**Pull request by jppelteret**
Relates to [this forum post](https://groups.google.com/forum/#!topic/dealii/19snsi6qJ_k).

### CODE REVIEW

File: include/deal.II/base/config.h.in [hunk 1]
```diff
@@ -138,6 +138,9 @@
  * cmake/modules/Find*.cmake
  */
 
+/* cmake/configure/configure_3_petsc.cmake */
```
Comment (tamiko) [inline include/deal.II/base/config.h.in]: Can you change the comment to `/* cmake/modules/FindPETSC.cmake */ and put it in alphabetical order? :-)

Comment (jppelteret) [reply include/deal.II/base/config.h.in]: But this is set in [`configure_3_petsc.cmake`](https://github.com/dealii/dealii/blob/master/cmake/configure/configure_3_petsc.cmake#L139)?

**Comment by tamiko** [comment]
> But this is set in configure_3_petsc.cmake

We do the renaming there. The logic is in the FindPETSC.cmake file.
(This is also true for the other two comments immediately following.)

**Comment by jppelteret** [comment]
@tamiko I've made the change but I must admit that I don't agree with your reasoning. We're basically directing someone to a file that doesn't define `DEAL_II_PETSC_WITH_MUMPS`, but another macro that is related to it (in this case `PETSC_WITH_MUMPS`). I don't think that's transparent for someone unfamiliar with the build system (i.e. me). If one points to the associated `configure` file, then at least one can see exactly what the macro is set to and then jump to the `Find` module to see how the internal/external definition is made/extracted.

**Comment by koecher** [comment]
I think the problematic line is to be found here in the cmake build system:
```
[koecheru@lillehammer deal.II-v8.5.1]$ find . -type f -exec grep -H 'DEAL_II_PETSC_WITH_MUMPS' {} \;
./cmake/configure/configure_3_petsc.cmake:SET(DEAL_II_PETSC_WITH_MUMPS ${PETSC_WITH_MUMPS})
```
it should be
```
./cmake/configure/configure_3_petsc.cmake:SET(DEAL_II_PETSC_WITH_MUMPS ${PETSC_HAVE_MUMPS})
```
A simple test for  `PETSC_HAVE_MUMPS`, `DEAL_II_PETSC_WITH_MUMPS`, `PETSC_WITH_MUMPS` results currently (deal.II v8.5.1 via candi) in `true`, `false`, `false`.

Petsc with mumps is installed, if the package "Parmetis" and "Petsc" is enabled in the candi configuration.

Kind regards
  Uwe

**Comment by jppelteret** [comment]
@koecher I don't think that thats the issue (although I could be mistaken). If you look at the [`_petsc_feature_check`](https://github.com/dealii/dealii/blob/master/cmake/modules/FindPETSC.cmake#L57) defined a few lines above where the line [defining the PETSc MUMPS regex](https://github.com/dealii/dealii/blob/master/cmake/modules/FindPETSC.cmake#L71) is, then you'll see that we transfer the result of that search to an internal variable `PETSC_WITH_MUMPS`.

**Comment by koecher** [comment]
@jppelteret Indeed, you are right, I've checked the line with manual output (without the proposed change)
```
SET(DEAL_II_PETSC_WITH_MUMPS ${PETSC_WITH_MUMPS})
```
and it is correctly set in the configure_3_petsc.cmake to true.
Hmpf, but anyhow this variable gets somehow not defined in the final installation.

**Comment by bangerth** [comment]
I'm not fundamentally opposed to this patch, given that we explicitly support MUMPS wrappers. But we can't promise to copy *every* PETSc configuration flag. We also don't need to: user codes can access them via `petscconf.h`.

**Comment by davydden** [comment]
> We also don't need to

not necessarily. We might eventually have some unit tests which rely on `petsc` having `mumps`  (i.e. filename being `foo.petsc_with_mumps.output`) and thereby it's good to have this patch.

**Comment by tamiko** [comment]
On second thought, I agree.
If we do not need `(DEAL_II_)PETSC_WITH_MUMPS` in the configuration - we should not export it.

What about refactoring the source file to include `petscconf.h` and get the information from there?

**Comment by tjhei** [comment]
> What about refactoring the source file to include petscconf.h and get the information from there?

While reasonable for the implementation, I can see that we want to write tests using ``petsc_with_mumps`` in the filename (hang on, we don't have a single test for it?!) as @davydden hinted at.

**Comment by bangerth** [comment]
But doesn't that only mean that we have to have this information available in cmake, but not in the deal.II `config.h`?

**Comment by davydden** [comment]
> hang on, we don't have a single test for it?!

yes, this part is not being tested at all.

**Comment by davydden** [comment]
> If we do not need (DEAL_II_)PETSC_WITH_MUMPS in the configuration - we should not export it.

If we have `DEAL_II_PETSC_WITH_MUMPS`, it's convenient for user projects to require that in `CmakeLists.txt` along with things like `DEAL_II_WITH_PETSC` etc.

So i personally don't see any inconsistencies in exporting this variable.

**Comment by jppelteret** [comment]
I think that @davydden has raised all of the points that I would have (and more). I remain of the opinion that this change is a sensible one, particularly when factoring in a mechanism to have users enable/disable the use of this solver in their own code. 

Tangential point: I was going to say "surely we do the same with `UMFPACK` and `sparse_direct`", but I would be half wrong - we don't have guards around the whole header/source, but [only for the includes](https://github.com/dealii/dealii/blob/master/include/deal.II/lac/sparse_direct.h#L29). Isn't this wrong? Can't we completely disable `UMFPACK`, or will the bundled version always be built if no external library is found? I must be missing something here...

**Comment by tjhei** [comment]
> surely we do the same with UMFPACK

We compile that code, always allow creating the sparse direct preconditioner, and throw an exception at runtime if deal.II was not installed with UMFPACK and you try to use it. I guess that makes sense, because you would need to litter the user code with ifdef's (think of an example where you can switch between different solvers using a .prm, so you need to be able to compile it even if you don't want to use it).

**Comment by jppelteret** [comment]
@tjhei Thanks for the explanation - that does make sense!

**Comment by drwells** [comment]
@jppelteret I see `#ifdef DEAL_II_WITH_UMFPACK` checks in `source/lac/sparse_direct.cc`; did you mean something else?

I support this patch too: honestly, I am not sure how we manage to currently pick up the MUMPS tests but things are completely unambiguous with this one (i.e., since `DEAL_II_WITH_PETSC_WITH_MUMPS` ends in `WITH_PETSC_WITH_MUMPS` the current test output file name `sparse_direct_mumps.with_petsc_with_mumps=on.with_64bit_indices=off.output` is clearly handled by the pickup test macro).

**Comment by drwells** [comment]
I have to admit that we are a little inconsistent here: the other two symbols we should `#define` wrappers for are `PETSC_HAVE_HYPRE` and `PETSC_HAVE_BLASLAPACK`. If we are in agreement regarding this patch then I would like to go ahead and add those so that we are completely covered. I agree that we should not wrap every single PETSc `#define` but I think wrapping these three definition checks is okay.

**Comment by davydden** [comment]
@bangerth another reasons we can use PETSC_WITH_MUMPS: https://github.com/dealii/dealii/issues/5141

**Comment by davydden** [comment]
`PETSC_HAVE_HYPRE` is useful indeed, not sure where we would befinit from `PETSC_HAVE_BLASLAPACK`.

**Comment by jppelteret** [comment]
>  we should #define wrappers for are PETSC_HAVE_HYPRE

Agreed

> I see #ifdef DEAL_II_WITH_UMFPACK checks in source/lac/sparse_direct.cc; did you mean something else?

@drwells You are quite right... I did mean `source/lac/sparse_direct.cc` but I completely missed [this line](https://github.com/dealii/dealii/blob/master/source/lac/sparse_direct.cc#L52) :-/ I'll get my glasses checked.

**Comment by drwells** [comment]
@davydden The PETSc documentation states that LAPACK is a required package, so, in fact, we should just assume that this is true. Now I am not sure when this preprocessor constant would ever not be defined; its not like the PETSc guys to leave things lying around for no reason.

**Comment by bangerth** [comment]
I have no objection to this patch. Does anyone?

/run-tests

**Comment by davydden** [comment]
@bangerth i don't have any objections, but i think we agreed above that we need the same for `hypre` in this PR

**Comment by bangerth** [comment]
That's ok with me, but it can also happen as a follow-up if getting this one merged would be useful.

**Comment by drwells** [comment]
Yes, I can add the hypre check later today.

At least five of us ( @davydden, @tamiko, @jppelteret, myself, and @bangerth) seem to approve of this patch so lets just merge it if the tester is happy.

**Comment by tjhei** [comment]
> Yes, I can add the hypre check later today.

I agree to this PR and adding a variable for HYPRE. This is mostly so we can conditionally run tests, I think it is less important for users.

**Pull request merged**